include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td

```diff
@@ -29,6 +29,16 @@ def TritonGPU_Dialect : Dialect {
             "TritonGPU module should contain a triton_gpu.num-warps attribute");
       return numWarps.cast<IntegerAttr>().getInt();
     }
+
+    static std::string getThreadsPerWarpAttrName() { return "triton_gpu.threads-per-warp"; }
+    static int getThreadsPerWarp(ModuleOp mod) {
+      Attribute threadsPerWarp = mod->getDiscardableAttr("triton_gpu.threads-per-warp");
+      if(!threadsPerWarp) {
+        return 32;
```

Comment:
Does this mean the default value is set to 32?
I'd suggest to set 64 as the default value, since MI series are more important targets.

Reply:
I'm keeping it untouched intentionally, otherwise we will have one extra conflict to deal with whenever sync with upstreams.

Additionally I don't think keeping the default be 32 to accommodate MI is a better approach because
1. The upstream is unlikely to accept 64 as default
2. The divergence makes the behavior of ROCM/triton different with upstream
3. New AMD GPU users need to take extra care NOT using the upstream otherwise their MI GPU system would potentially fail on Triton due to the divergence.

To avoid this we need to make sure Triton works on MI with 32 as default at official branch, rather than forcing potential new users seeking for solutions from some fork. Setting default to 64 only hides potential problems away from users.

Reply:
I dont think MI gpus work with warpSize = 32 when it comes mfma instructions. 
yes, in terms of compatibility, we dont want to diverge from upstream. And some day we also want to live with nvidia gpus in upstream. 
But for now, most of critical applications are running on MI gpu. And a default value of 32 means that we need to change for those applications. 
@jayfurmanek what do you think?

Reply:
The whole point of the PR is to make sure the warp size can change without re-compiling Triton package, and theoretically on a system with both Navi 31 and MI200 installed, we can run Triton on both GPUs at the same time without installing two copies of Triton.

If MI failed with warp size defaults to 32, it means the Triton JIT compiler has problems figuring out the right warp size for the target architecture.

Reply:
>>To avoid this we need to make sure Triton works on MI with 32 as default at official branch
@xinyazhang Do you mean let 32 be the default but do a dynamic check/set based on GPU type?

It needs to be 64 for the MI GPUs at the end of the day.

Reply:
Another solution is to make the default warp size be a compile time macro and configurable at compile time.

Reply:
> Do you mean let 32 be the default but do a dynamic check/set based on GPU type?

Yes, this is the whole point of this PR.
The detection logic is in `./python/triton/compiler/compiler.py`. Ultimately it calls `hipGetDeviceProperties` and passes the warpSize to the constructor of the Triton JIT compiler.

Reply:
ok - I'll look through it closer. Is there even a scenario where the default is used? Maybe only if the arch is unrecognized?

Reply:
> ok - I'll look through it closer. Is there even a scenario where the default is used? Maybe only if the arch is unrecognized?

A tl;dr version is the default value should never be used at runtime., but the story about warp size is a little long. 

1. We want the warpSize to be a dynamical property, and hence it should be a property specific to some MLIR pass(es)    
2. This demands leads to the **UPSTREAM** code change at [`include/triton/Conversion/TritonToTritonGPU/Passes.td`](https://github.com/ROCmSoftwarePlatform/triton/pull/244/files#diff-e219b4a9f8bdfbd57b2134863508395eeed113eb9fec36811d055aed9938f68b), which create a new property called threadsPerWarp
3. However there is no direct way to change the property programmatically. The only documented way is using command line options but we don't use it in that way
4. In practice, `class ConvertTritonToTritonGPU` (at file `lib/Conversion/TritonToTritonGPU/TritonToTritonGPUPass.cpp`) inherits `ConvertTritonToTritonGPUBase<ConvertTritonToTritonGPU>` and properly configures the warpSize in its constructor. This constructor is invoked by `createConvertTritonToTritonGPUPass` which is exported as `add_convert_triton_to_tritongpu_pass` in python. This python interface is called by `compiler.py` according to the warpSize queried from `hipGetDeviceProperties`
5. However, there is another function `createConvertTritonToTritonGPUPass()` which doesn't take any argument and creates an `ConvertTritonToTritonGPU` object with default settings, leading to errors. This function should never be called at runtime, but I cannot remove it because this function is required in `registerConvertTritonToTritonGPU()` (generated as `./python/build/cmake.linux-x86_64-cpython-3.10/include/triton/Conversion/TritonToTritonGPU/Passes.h.inc`). My knowledge about MLIR is not enough to remove this entrance potentially in danger.
6. Other properties are fundamentally propagated from `ConvertTritonToTritonGPU` object. An example propagation path is `threadsPerWarp = triton::gpu::TritonGPUDialect::getThreadsPerWarp(mod)->decomposeMmaToDotOperand(..., decomposeMmaToDotOperand)->layout = triton::gpu::BlockedEncodingAttr::get(..., threadsPerWarp)->...->getWarpSize(layout)`

Reply:
Thanks @xinyazhang for such a nice clarification about warpSize.
Now I see where it comes from:
- When the jit compiler is invoked, i.e. `python ./some_test.py`, warpSize is determined by arch, which is determined by hipGetDeviceProperties. In this case, we do not need to set warpSize.
- When triton-opt is used to lower IRs, warpSize can be set as the parameter. E.g. `triton-opt –convert-triton-to-tritongpu=threadsPerWarp=64`. If this parameter is not set, the default value is 32 in this PR and also upstream

Both paths will go through `ConvertTritonToTritonGPU` and set `triton_gpu.threads-per-warp` as an attribute of the module. Therefore, later passes can call `triton::gpu::TritonGPUDialect::getThreadsPerWarp(mod)` to obtain warpSize.
I’d argue that we change this default value to 64 in favor of MI gpus (at least for now) so that when we use triton-opt on MI gpus,
- we do not need to set the warpSize explicitly as the parameter
- we do not need to include triton_gpu.threads-per-warp in the module. Therefore, we do not need to change existing IRs

Reply:
> * we do not need to set the warpSize explicitly as the parameter
> * we do not need to include triton_gpu.threads-per-warp in the module. Therefore, we do not need to change existing IRs

Once again this PR is for Navi 2x/3x support, and both changes are needed unless there are better propagation approaches.
If MI is considered as the only target that need to take care of I can then close the PR.

Reply:
I think this PR does a great job in enabling Navi gpus. And I also agree that detecting arch at runtime and set warpSize accordingly is an elegant solution to the problem. From this angle, the default value of `threadsPerWarp` does not make any difference.

However, what I am trying to bring up is the scenario where `triton-opt` is used. And the default value does make a difference here.

> Once again this PR is for Navi 2x/3x support, and both changes are needed unless there are better propagation approaches.

Both changes are needed only on MI gpus.

> If MI is considered as the only target that needs to take care of I can then close the PR.

Changing the default to 64 does not mean MI is the only target. Just as keeping it as 32 does not mean Navi is the only target.

> Update: oh I forgot about one more thing. All changes in https://github.com/ROCmSoftwarePlatform/triton/commit/b95482d9ff298b3c5bfaba3677331499dd1fb9c6 come from the Triton upstream directly. I'm not the right person to argue with.

NVIDIA gpus are the main hw for upstream, hence 32 is a valid choice as the default value. So nobody is going to argue that with upstream people.